---
title: "4. Analyze Variant Patterns"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

In this notebook, I'll take the merged and filtered variants and look for patterns that emerge. This includes:
  - How the frequency and distribution of variants changes conditions?
  - How the frequency and distribution of variants changes over time?
  - Are there shared variants between multiple samples or cages?
  - How do the number of variants change by viral copy number? 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "ggrepel", "data.table")
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {
  
  # Sample metadata
  sample.metadata.data = snakemake@input[[1]]
  
  # Variant calls from pysam, varscan, ivar, and lofreq
  variant.calls.data = snakemake@input[[2]]
  
  # Merged variant calls from the previous notebook 
  merged.variant.calls.data = snakemake@input[[3]]
    
  # Primer BED
  primer.bed.data = snakemake@input[[4]]

} else {

  # Sample metadata
  sample.metadata.data = "../../config/samples.csv"
  
  # Variant calls from pysam, varscan, ivar, and lofreq
  variant.calls.data = "../../results/variants/variants.csv"
  
  # Merged variant calls from the previous notebook 
  merged.variant.calls.data = "../../results/variants/merged_variants.csv"
  
  # Primer BED
  primer.bed.data = "../../results/reference/primers/primers.bed"
  
}

```

```{r Output Directory, echo=T}

figures.dir = "../../results/figures/"

```

```{r Load Data, echo=T}

merged.df = read_csv(merged.variant.calls.data, show_col_types = FALSE) %>% 
  # Fix the naming scheme of experiment 19 
  mutate(Condition = case_when(
    Experiment == 19 & Pair == 'aerosol' ~ 'aerosol transmission',
    Experiment == 19 & Pair == 'direct' ~ 'direct transmission',
    TRUE ~ Condition
  )) %>% 
  # Standardize the pair names to contact and donor
  mutate(Pair = case_when(
    Pair == 'aerosol' ~ 'contact',
    Pair == 'direct' ~ 'contact',
    Pair == 'recipient' ~ 'contact',
    TRUE ~ Pair
  ))

# Primer positions
primer.df = fread(primer.bed.data) %>% 
  select(contig = V1,
         start = V2, 
         end = V3, 
         name = V4, 
         qual = V5, 
         direction = V6)
```

## Allele Frequency Distribution

Let's see what the distribution of allele frequencies are across all of our samples. This includes the isolates and fixed variants. 

```{r Frequency Distribution, fig.align='center', fig.width=6, fig.height=6}

merged.df %>% 
  ggplot(aes(x = AF)) + 
    geom_histogram(bins = 30) +
    xlab("Allele Frequency") + 
    ylab("Allele Count") + 
    theme_bw(18)

ggsave(paste(figures.dir, "frequency_distribution.png", sep=""), width = 6, height = 6, dpi = 300)

```

But, it might be more informative to filter out the variants that are fixed in the isolate. Let's see what this does. 

```{r Fixed Stock Variants, echo=T}

fixed.isolate.variants = merged.df %>%
  filter(Animal == "stock") %>% 
  filter(AF >= .99) %>% 
  pull(SNP)

```

In total, there are `r length(fixed.isolate.variants)` fixed in the stock that was used to inoculate the donors. 

```{r Frequency Distribution no Fixed, fig.align='center', fig.width=6, fig.height=6}

merged.df %>% 
  filter(!SNP %in% fixed.isolate.variants) %>% 
  ggplot(aes(x = AF)) + 
    geom_histogram(bins = 30) +
    xlab("Allele Frequency") + 
    ylab("Allele Count") + 
    theme_bw(18)

ggsave(paste(figures.dir, "frequency_distribution_no_fixed_isolate_SNPs.png", sep=""), width = 6, height = 6, dpi = 300)

```

We can see that there are still some fixed variants. There are likely variants that fixed either during transmission or over the course of the infection. Let's get a better idea of this by plotting the distribution in frequency by condition, specifically, whether the alleles came from the donor or the recipient. 

```{r Frequency Distribution Donor v. Recip, fig.align='center', fig.width=12, fig.height=6}

merged.df %>% 
  filter(!is.na(Pair)) %>% 
  filter(!SNP %in% fixed.isolate.variants) %>% 
  ggplot(aes(x = AF, fill = factor(Pair, levels = c("donor", "contact")))) + 
    geom_histogram(bins = 30) +
    facet_wrap(~factor(Pair, levels = c("donor", "contact")), ncol=2) + 
    xlab("Allele Frequency") + 
    ylab("Allele Count") + 
    scale_fill_manual(values = c("darkblue", "darkred"), name="") + 
    theme_bw(18) + 
    theme(legend.position = "bottom")

ggsave(paste(figures.dir, "frequency_distribution_by_pair.png", sep=""), width = 12, height = 6, dpi = 300)

```

Indeed, it seems to be the case the variants are higher frequency, even fixed, after transmission occurs. This is pretty neat, but one caveat is that time after infection could play a role. In other words, do variants tend to fix over time, or is a bottleneck at transmission fixing the variants? 

```{r Frequency Distribution Time, fig.align='center', fig.width=18, fig.height=6}

merged.df %>% 
  filter(!SNP %in% fixed.isolate.variants,
         Animal != "stock",
         Pair == "contact") %>% 
  mutate(DPC = paste(DPC, "Days Post Contact")) %>% 
  ggplot(aes(x = AF)) + 
    geom_histogram(bins = 30) +
    facet_wrap(~DPC, ncol=3) + 
    xlab("Allele Frequency") + 
    ylab("Allele Count") + 
    theme_bw(18) + 
    theme(legend.position = "bottom") 

ggsave(paste(figures.dir, "frequency_distribution_by_time.png", sep=""), width = 18, height = 6, dpi = 300)

```

Generally, there appear to be fewer low frequency variants relative to fixed variants over time, but this is also likely a function of depth and viral copy number limiting the resolution of variant calling in the later time points. Also, the pattern of fixed and high frequency variants after transmission is plainly apparent in the earliest time point. 

## Variant Resolution

To be sure, let's check how the number of variants (excluding variants fixed in the stock) varies as a function of viral copy number and TCID50. 

```{r Number of Variants v. Copies, fig.align='center', fig.width=6, fig.height=6}

merged.df %>% 
  filter(!SNP %in% fixed.isolate.variants,
         Animal != "stock") %>% 
  group_by(Animal, DPC, TCID50, RNA_Copies) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(Label= paste(Animal, " ", DPC, "-DPC", sep="")) %>% 
  ggplot(aes(x = RNA_Copies, y = n, col = DPC, label = Label)) + 
    geom_point() + 
    geom_text_repel(size = 2) + 
    ggtitle("SNPs vs. RNA Copies") +
    xlab("RNA Copies") + 
    ylab("Allele Count") + 
    theme_bw(18)+ 
    theme(legend.position = "none",
          plot.title = element_text(hjust = 0.5))
    
ggsave(paste(figures.dir, "variants_v_RNA_Copies.png", sep=""), width = 6, height = 6, dpi = 300)

```

```{r Number of Variants v. TCID50, fig.align='center', fig.width=6, fig.height=6}

merged.df %>% 
  filter(!SNP %in% fixed.isolate.variants,
         Animal != "stock") %>% 
  group_by(Animal, DPC, TCID50, RNA_Copies) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(Label= paste(Animal, " ", DPC, "-DPC", sep="")) %>% 
  ggplot(aes(x = TCID50, y = n, col = DPC, label = Label)) + 
    geom_point() + 
    geom_text_repel(size = 2) + 
    ggtitle("SNPs vs.TCID50") +
    xlab("TCID50") + 
    ylab("Allele Count") + 
    theme_bw(18)+ 
    theme(legend.position = "none",
          plot.title = element_text(hjust = 0.5))
    
ggsave(paste(figures.dir, "variants_v_TCID50.png", sep=""), width = 6, height = 6, dpi = 300)

```

Interestingly, although RNA Copies seems to have a slight positive correlation with the number of variants called in a given sample, the TCID50 is a much better predictor. 

## Shared Variants

In the next analysis, I'll take these variants and attempt to calculate the transmission bottleneck for different sizes of aerosol and modes of transmission. However, this analysis can be heavily impacted by false positive variants. One possible indicator of false positive variants are variants shared between many unrelated samples. I'll investigate whether there are such variants present in our data below.

First, I'll plot the distribution of variants shared across samples. 

```{r All Shared Variants, warning = F, fig.align='center', fig.width=6, fig.height=6}

number.of.conditions = merged.df %>% select(Animal, DPI) %>% distinct() %>% nrow()

merged.df %>% 
  select(SNP, Animal, DPC) %>% 
  group_by(SNP) %>% 
  count() %>% 
  ggplot(aes(x = n)) + 
    geom_histogram(bins = 30) + 
    xlab("Number of Samples") +
    ylab("Number of Alleles") +
    scale_x_continuous(limits = c(0, number.of.conditions)) + 
    theme_bw(18)

ggsave(paste(figures.dir, "variants_v_num_samples.png", sep=""), width = 6, height = 6, dpi = 300)

```

As expected, although the majority of variants are unique, many are shared between samples. Some appear common to every sample. It's likely that a significant fraction of these are variants that were present in the isolate. Let's filter these out and see how the distribution changes. 

```{r Isolate Dropped Shared Variants, warning = F, fig.align='center', fig.width=6, fig.height=6}

isolate.variants = merged.df %>% 
  filter(Animal == "stock") %>% 
  pull(SNP)

merged.df %>% 
  filter(Animal != "stock",
         !SNP %in% isolate.variants
         ) %>% 
  select(SNP, Animal, DPC) %>% 
  distinct() %>% 
  group_by(SNP) %>% 
  count() %>% 
  ggplot(aes(x = n)) + 
    geom_histogram(bins = 30) + 
    xlab("Number of Samples") +
    ylab("Number of Alleles") +
    scale_x_continuous(limits = c(0, number.of.conditions)) + 
    theme_bw(18)

ggsave(paste(figures.dir, "variants_v_num_samples_no_isolate.png", sep=""), width = 6, height = 6, dpi = 300)

```

This certainly improves the distribution. The issue with filtering out variants that are present in the isolate is that although many of these could be real, the same things that lead to false-positives in multiple samples will likely also affect the variants called in the isolate. 

## Possible False Positives

The key indication of a possible false-positive variant is that it's present in more than one cage. If a variant is present in more than one animal, that could be the results of transmission. If a variant is present in more than one cage, there are three possible options: 

1. **It's a real variant** present in the stock that was used to inoculate the donor animals. 
2. **It's a real variant** that occurred multiple times due to selection.
3. **It's a false variant** that's a recurrent artifact.

Below, I'll try to isolate the artifacts and determine what's special about these variants. If there are unique features of these possible false-positives, we can apply filters that remove these artifacts. 

```{r Shared Variants Across Experiments, warning = F, fig.align='center', fig.width=6, fig.height=6}

merged.df %>% 
  filter(Animal != "stock",
         !SNP %in% fixed.isolate.variants
         ) %>% 
  select(SNP, Experiment) %>% 
  distinct() %>% 
  group_by(SNP) %>% 
  count() %>% 
  ggplot(aes(x = n)) + 
    geom_histogram(bins = 30) + 
    xlab("Number of Experiments") +
    ylab("Number of Alleles") +
    theme_bw(18)

ggsave(paste(figures.dir, "variants_v_cages.png", sep = ""), width = 6, height = 6, dpi = 300)

```

Although the vast number of variants are unique to a single cage, there are many that are present in multiple *if not all cages*. As a note, I'm not including the variants that are fixed in the isolate. These are variants we expect to be present in all cages unless there is a reversion or some other mutation at this position. 

```{r Variants in Multiple Cages, echo=T}

# Get the SNPs that arose in more than one cage
multi.cage.variants = merged.df %>% 
  filter(Animal != "stock",
         !SNP %in% fixed.isolate.variants
         ) %>% 
  select(SNP, Experiment) %>% 
  distinct() %>% 
  group_by(SNP) %>% 
  count() %>% 
  filter(n > 1) %>% 
  pull(SNP) 

# Get the SNPs that are in primer stretches 
snps.in.primers = merged.df %>% 
  select(POS, SNP) %>% 
  left_join(., select(primer.df, c(start, end)), by = character()) %>% 
  mutate(in_primer = between(POS, start, end, incbounds = TRUE)) %>% 
  filter(in_primer) %>% 
  pull(SNP) %>% 
  unique()

# Annotate the merged variants with signs of artifacts
artifact.investigation.df = merged.df %>% 
  filter(!SNP %in% fixed.isolate.variants) %>% 
  mutate(Multiple_Cages = if_else(SNP %in% multi.cage.variants, T, F),
         OBSV = ADR + ADF,
         BIAS = pmax(ADR, ADF) / OBSV,
         In_Primer = if_else(SNP %in% snps.in.primers, T, F))

```

There are `r length(multi.cage.variants)` present in multiple cages. Note that some of these will be real variants as per the examples highlighted above. Despite that, I'd guess that the majority of these variants will be false-positives. 

```{r Strand-Bias, fig.align='center', fig.width=6, fig.height=6}

mean.bais = artifact.investigation.df %>% 
  group_by(SNP, Multiple_Cages) %>% 
  summarize(BIAS = mean(BIAS)) %>% 
  ungroup() %>% 
  group_by(Multiple_Cages) %>% 
  summarize(BIAS = mean(BIAS))

artifact.investigation.df %>% 
  group_by(SNP, Multiple_Cages) %>% 
  summarize(BIAS = mean(BIAS)) %>% 
  ggplot(aes(x = Multiple_Cages, y = BIAS)) + 
    geom_jitter() + 
    geom_errorbar(data = mean.bais, aes(ymin = BIAS, ymax = BIAS), col = 'red') +
    ylab("Stand-Bias") + 
    xlab("Detected In Multiple Cages") + 
    theme_bw(18)

ggsave(paste(figures.dir, "strand_bias_shared_variants.png", sep = ""), width = 6, height = 6, dpi = 300)

```

There is certainly a higher than average strand-bias in the variants that are shared between multiple cages. Therefore, applying a conservative stand-bias filter would probably help remove false-positives. 

```{r Observations, fig.align='center', fig.width=6, fig.height=6}

mean.OBSV = artifact.investigation.df%>% 
  group_by(SNP, Multiple_Cages) %>% 
  summarize(OBSV = mean(OBSV)) %>% 
  ungroup() %>% 
  group_by(Multiple_Cages) %>% 
  summarize(OBSV = mean(OBSV))

artifact.investigation.df %>% 
  group_by(SNP, Multiple_Cages) %>% 
  summarize(OBSV = mean(OBSV)) %>%  
  ggplot(aes(x = Multiple_Cages, y = OBSV)) + 
    geom_jitter() + 
    geom_errorbar(data = mean.OBSV, aes(ymin = OBSV, ymax = OBSV), col = 'red') +
    ylab("Number of Variant Reads") + 
    xlab("Detected In Multiple Cages") + 
    theme_bw(18)

ggsave(paste(figures.dir, "obsv_count_shared_variants.png", sep = ""), width = 6, height = 6, dpi = 300)

```

I think that this is slightly less interesting, but the average number of reads with a variant allele is lower in the possible false positives. 

```{r Primer Overlaps, fig.align='center', fig.width=6, fig.height=6}

artifact.investigation.df %>% 
  select(Multiple_Cages, In_Primer, SNP) %>% 
  distinct() %>% 
  group_by(Multiple_Cages, In_Primer) %>% 
  count() %>% 
  group_by(Multiple_Cages) %>% 
  mutate(Prop = n / sum(n)) %>% 
  ggplot(aes(x = Multiple_Cages, y = Prop, fill = In_Primer)) + 
    geom_bar(position = position_stack(), stat = "identity") + 
    scale_fill_manual(values = c("black", "darkred"), name = "Overlaps Primer:") +
    ylab("Proportion of SNPs") + 
    xlab("Detected In Multiple Cages") + 
    theme_bw(18) +
    theme(legend.position = "bottom")

ggsave(paste(figures.dir, "primer_overlap_shared_variants.png", sep = ""), width = 6, height = 6, dpi = 300)

```

Additionally, the many of the variants detected in multiple cages overlap with primer sequences. This is in spite of the primer trimming that I preformed in the pre-processing step. 

*Of note, for plots that involve quantitative metrics, I'm using the average across all shared SNPs.* 

