---
title: "2. Compare Variant Callers"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

In this notebook, I compare the overlap in variants called by the four variant calling approaches that I used: 

- `iVar`
- `Varscan`
- `Lofreq`
- `Pysam`

I'll determine a set of variants to use in the downstream analysis. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "UpSetR", "data.table")
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {
  
  # Depth from `samtools depth` after primer trimming
  trimmed.depth.data = snakemake@input[[2]]
  
  # Sample metadata
  sample.metadata.data = snakemake@input[[3]]
  
  # Variant calls from pysam, varscan, ivar, and lofreq
  variant.calls.data = snakemake@input[[4]]
  
  
} else {

  # Depth from `samtools depth` after primer trimming
  trimmed.depth.data = "../../results/coverage/merged.trimmed.depth.tsv"
  
  # Sample metadata
  sample.metadata.data = "../../config/samples.csv"
  
  # Variant calls from pysam, varscan, ivar, and lofreq
  variant.calls.data = "../../results/variants/variants.csv"
  
}

```

## Variant Overlap 

Below, I'll see what the overlap in variant calls are for the four approaches (`iVar`, `Varscan`, `Lofreq`, and `Pysam`). I did my best to standardize a minimal set of filters across every approach. To see these filters, look in the `config/snakemake.yml` file. 

```{r Load Variants, echo=TRUE, warning=FALSE, message=FALSE}

variants.df = read_csv(variant.calls.data, show_col_types = FALSE) %>% 
    mutate(SNP = paste(REF, POS, ALT, sep = "")) 

```

The best way to visualize the overlap between variant callers is an `UpSet` plot. Below, I'll plot the overlap of variant callers without any extra filters applied.

```{r Raw Variants Overlap, fig.align='center', fig.width=6, fig.height=3}

variants.df %>% 
  select(Run, SNP, CALLER) %>% 
  mutate(Identifier = paste(Run, SNP, sep = "-")) %>% 
  select(Identifier, CALLER) %>% 
  distinct() %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "CALLER", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., 
        sets = c("lofreq", "varscan", "ivar", "pysam"),
        sets.bar.color = "#56B4E9",
        order.by = "freq",
        empty.intersections = "on",
        text.scale = 1.5)

```

A significant number of the total variants are shared between all callers even without filters applied. The two callers with the most unique variants are `lofreq` and `pysam`. It's not surprising that `pysam` has many unique variants given that there are no probability based filters or tests applied. It's a bit more surprising that `lofreq` has so many variants, however, it's known to be a very sensitive variant caller that can have many false positives. 

Now, I'll apply some basic thresholds to see if I can improve the overlap. In theory, the greater overlap between approaches, the less change we're identifying false-positive variants. 

```{r Variant Calling Thresholds, echo=T}

min.freq = 0.02
min.depth = 200
min.obsv = min.depth * min.freq

```

Below, I'm using a minimum allele frequency of `r min.freq`, a minimum depth of `r min.depth`, a minimum number of variants reads of `r min.obsv`, and a maximum strand bias of `r paste((strand.bias * 100), "%", sep="")` variants reads on a single strand.

```{r Filtered Variant Overlap, fig.align='center', fig.width=6, fig.height=3}

filtered.variants.df = variants.df %>% 
  filter(
    AF >= min.freq,
    DP >= min.depth,
    ADF + ADR >= min.obsv
    )

filtered.variants.df %>% 
  select(Run, SNP, CALLER) %>% 
  mutate(Identifier = paste(Run, SNP, sep = "-")) %>% 
  select(Identifier, CALLER) %>% 
  distinct() %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "CALLER", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., 
        sets = c("lofreq", "varscan", "ivar", "pysam"),
        sets.bar.color = "#56B4E9",
        order.by = "freq",
        empty.intersections = "on",
        text.scale = 1.5)

```

This drastically improves the overlap between the callers. Now, essentially, the only unique variants are called with `lofreq`. To come up with a set of variants to continue with, I'll take the variant alleles called by three or more variant callers. 

It's import to also see how this affects the correlation between replicates as well. I think a general approach that could would is to keep all of the variants that are present in `Pysam`, `Varscan`, and `iVar`. Since there are so many variants that are unique to `lofreq` and this variant caller is known to be sensitive, but prone to false positives, those variants should probably be ignored. 

In the next notebook I'll see how the correlation between replicates look between callers anc combine the variants. 


