---
title: "2. Compare Variant Callers"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

In this notebook, I compare the overlap in variants called by the four variant calling approaches that I used: 

- `iVar`
- `Varscan`
- `Lofreq`
- `Pysam` (A custom approach written in `Python`)

I'll determine a set of variants to use in the downstream analysis. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "UpSetR", "data.table")
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {
  
  # Depth from `samtools depth` after primer trimming
  trimmed.depth.data = snakemake@input[[1]]
  
  # Sample metadata
  sample.metadata.data = snakemake@input[[2]]
  
  # Variant calls from pysam, varscan, ivar, and lofreq
  variant.calls.data = snakemake@input[[3]]
  
  
} else {

  # Depth from `samtools depth` after primer trimming
  trimmed.depth.data = "../../results/coverage/merged.trimmed.depth.tsv"
  
  # Sample metadata
  sample.metadata.data = "../../config/samples.csv"
  
  # Variant calls from pysam, varscan, ivar, and lofreq
  variant.calls.data = "../../results/variants/variants.csv"
  
}

```

```{r Output Directory, echo=T}

figures.dir = "../../results/figures/"

```

## Variant Overlap 

Below, I'll see what the overlap in variant calls are for the four approaches (`iVar`, `Varscan`, `Lofreq`, and `Pysam`). I did my best to standardize a minimal set of filters across every approach. To see these filters, look in the `./config/snakemake.yml` file at the root of the directory.

```{r Load Variants, echo=TRUE, warning=FALSE, message=FALSE}

variants.df = read_csv(variant.calls.data, show_col_types = FALSE) %>% 
    mutate(SNP = paste(REF, POS, ALT, sep = "")) 

```

The best way to visualize the overlap between variant callers is an `UpSet` plot. Below, I'll plot the overlap of variant callers without any extra filters applied.

```{r Raw Variants Overlap, message=F, fig.align='center', fig.width=8, fig.height=5}

raw.upset = variants.df %>% 
  select(Run, SNP, CALLER) %>% 
  mutate(Identifier = paste(Run, SNP, sep = "-")) %>% 
  select(Identifier, CALLER) %>% 
  distinct() %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "CALLER", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., 
        sets = c("lofreq", "varscan", "ivar", "pysam"),
        sets.bar.color = "#56B4E9",
        order.by = "freq",
        empty.intersections = "on",
        text.scale = 1.5)

raw.upset

```

```{r Save Raw Upset, echo=FALSE, message=F}

png(file=paste(figures.dir, "raw_caller_upset_plot.png", sep=""), width=8, height=5, units="in", res = 300)
raw.upset
dev.off()

```

A significant number of the total variants are shared between all callers even without filters applied. The two callers with the most unique variants are `lofreq` and `pysam`. It's not surprising that `pysam` has many unique variants given that there are no probability based filters or tests applied. It's a bit more surprising that `lofreq` has so many variants, however, it's known to be a very sensitive variant caller that can have many false-positives. 

Now, I'll apply some basic thresholds to see if I can improve the overlap. The greater overlap between approaches, the less likely it is that we've identified approach specific false-positive variants. 

```{r Variant Calling Thresholds, echo=T}

min.freq = 0.02
min.depth = 200
min.obsv = min.depth * min.freq

```

Below, I'm using a minimum allele frequency of `r min.freq`, a minimum depth of `r min.depth`, a minimum number of variants reads of `r min.obsv` variants reads on a single strand.

```{r Filtered Variant Overlap, message=F, fig.align='center', fig.width=8, fig.height=5}

filtered.variants.df = variants.df %>% 
  filter(
    AF >= min.freq,
    DP >= min.depth,
    ADF + ADR >= min.obsv
    )

filtered.upset = filtered.variants.df %>% 
  select(Run, SNP, CALLER) %>% 
  mutate(Identifier = paste(Run, SNP, sep = "-")) %>% 
  select(Identifier, CALLER) %>% 
  distinct() %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "CALLER", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., 
        sets = c("lofreq", "varscan", "ivar", "pysam"),
        sets.bar.color = "#56B4E9",
        order.by = "freq",
        empty.intersections = "on",
        text.scale = 1.5)

filtered.upset

```

```{r Save Filtered Upset, echo=FALSE, message=F}

png(file=paste(figures.dir, "filtered_caller_upset_plot.png", sep=""), width=8, height=5, units="in", res = 300)
filtered.upset
dev.off()

```

This drastically improves the overlap between the callers. Now, essentially, the only unique variants are called with `lofreq`. 

In the next notebook, I'll see how the correlation between replicates look between callers and combine the variants. 

## Paper Figures 

Below is the figure that will be included in the paper. It's identical to the plots above, but I'm excluding the samples that aren't included in the paper (aerosol and direct contact samples).


```{r Supplemental Figure 2, message=F, fig.align='center', fig.width=8, fig.height=5}

filtered.variants.df = variants.df %>% 
  filter(
    AF >= min.freq,
    DP >= min.depth,
    ADF + ADR >= min.obsv
    )

supplemental.figure.upset = filtered.variants.df %>% 
  select(Run, SNP, CALLER) %>% 
  mutate(Identifier = paste(Run, SNP, sep = "-")) %>% 
  select(Identifier, CALLER) %>% 
  distinct() %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "CALLER", values_from = "Tally", values_fill = 0) %>% 
  as.data.frame(row.names = Identifier) %>%
  upset(., 
        sets = c("lofreq", "varscan", "ivar", "pysam"),
        sets.bar.color = "#56B4E9",
        order.by = "freq",
        empty.intersections = "on",
        text.scale = 1.5)

filtered.upset

```

```{r Save Supplemental Figure 2, echo=FALSE, message=F}

png(file=paste(figures.dir, "supplemental-figre-2.png", sep=""), width=8, height=5, units="in", res = 300)
supplemental.figure.upset
dev.off()

```
