---
title: "6. Variants Over Time"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is to see what happens to transmitted variants over time. Is there any evidence of selection? How stable are variants over time? 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "data.table", "foreach")
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {
  
  # Sample metadata
  sample.metadata.data = snakemake@input[[1]]
  
  # Variant calls from pysam, varscan, ivar, and lofreq
  variant.calls.data = snakemake@input[[2]]
  
  # Merged variant calls from the previous notebook 
  merged.variant.calls.data = snakemake@input[[3]]
  
  # Primer BED
  primer.bed.data = snakemake@input[[4]]

} else {

  # Sample metadata
  sample.metadata.data = "../../config/samples.csv"
  
  # Variant calls from pysam, varscan, ivar, and lofreq
  variant.calls.data = "../../results/variants/variants.csv"
  
  # Merged variant calls from the previous notebook 
  merged.variant.calls.data = "../../results/variants/merged_variants.csv"
  
  # Primer BED
  primer.bed.data = "../../results/reference/primers/primers.bed"
  
}

```

```{r Output Directory, echo=T}

figures.dir = "../../results/figures/"

```

```{r Load Data, echo=TRUE, warning=FALSE}

merged.df = read_csv(merged.variant.calls.data, show_col_types = FALSE) %>% 
  # Fix the naming scheme of experiment 19 
  mutate(Condition = case_when(
    Experiment == 19 & Pair == 'aerosol' ~ 'aerosol transmission',
    Experiment == 19 & Pair == 'direct' ~ 'direct transmission',
    TRUE ~ Condition
  )) %>% 
  # Standardize the pair names to contact and donor
  mutate(Pair = case_when(
    Pair == 'aerosol' ~ 'contact',
    Pair == 'direct' ~ 'contact',
    Pair == 'recipient' ~ 'contact',
    TRUE ~ Pair
  ))

# Un-merged variant calls
variant.df = read_csv(variant.calls.data, show_col_types = FALSE)

# Primer positions
primer.df = fread(primer.bed.data) %>% 
  select(contig = V1,
         start = V2, 
         end = V3, 
         name = V4, 
         qual = V5, 
         direction = V6)

```

I'll use the filters that we determine in the previous notebook.

```{r Filter Variants, echo=TRUE}

# Try some more stringent filters
min.obsv = 25
min.bias = .9
min.freq = 0.02

# Get the SNPs that are in primer stretches 
snps.in.primers = merged.df %>% 
  select(POS, SNP) %>% 
  left_join(., select(primer.df, c(start, end)), by = character()) %>% 
  mutate(in_primer = between(POS, start, end, incbounds = TRUE)) %>% 
  filter(in_primer) %>% 
  pull(SNP) %>% 
  unique()
  
# Filter the merged data frame
filtered.merged.df = merged.df %>% 
  mutate(OBSV = ADR + ADF) %>% 
  mutate(BIAS = pmax(ADR, ADF) / OBSV) %>% 
  filter(
    OBSV >= min.obsv,
    BIAS <= min.bias,
    AF >= min.freq,
    !SNP %in% snps.in.primers
  )

```

In addition to the filters that were used to create the `merged.df`, I've increased the allele frequency filter to >= `r min.freq * 100`% , I've added a strand-bias filter of <= `r min.bias * 100`% variants on a single strand, and I've increased the minimum number of variant reads to `r min.obsv`. 

## Variant Stability

```{r Format Variant Data, echo = T, warning=FALSE}

# Get the amino acid annotations for each variant from Pysam
amino_acid_annotations = variant.df %>% 
  filter(CALLER == 'pysam') %>% 
  select(POS, REF, ALT, POS_AA, REF_AA, ALT_AA, GENE) %>% 
  mutate(SNP = paste(REF, POS, ALT, sep = "")) %>% 
  select(!c(POS, REF, ALT)) %>% 
  distinct()

# Fixed in the isolate
fixed.in.isolate = merged.df %>%  
  filter(Animal == "stock",
         AF >= .98) %>% 
  pull(SNP)
  
# Get the unique experiment names for cross joining with stock variants
faceting_columns = variant.df %>% 
  select(Experiment, Animal) %>% 
  filter(!is.na(Experiment)) %>% 
  distinct()

# Get all of the variants in the stock and expand by experiment (cross join)
stock.variants = variant.df %>% 
  filter(CALLER == 'pysam',
         Animal == "stock") %>% 
  mutate(DPC = -1,
         DPI = 0,
         Pair = "stock") %>% 
  select(POS, REF, ALT, POS_AA, REF_AA, ALT_AA, GENE, Pair, Condition, DPC, DPI, AF) %>% 
  left_join(., faceting_columns, by = character()) %>% 
  mutate(SNP = paste(REF, POS, ALT, sep = ""),
         Stock = T
         ) 

# Fix the fact that the same donor was used for aerosol and direct conditions
ivc.cage.variants = filtered.merged.df %>% 
  filter(Condition == "no impactor/IVC cage") 
ivc.cage.variants.aerosol = ivc.cage.variants %>% 
  mutate(Condition = "aerosol transmission")
ivc.cage.variants.direct = ivc.cage.variants %>% 
  mutate(Condition = "direct transmission")
amended.merged.df = filtered.merged.df %>% 
  filter(Condition != "no impactor/IVC cage") %>% 
  bind_rows(ivc.cage.variants.direct, ivc.cage.variants.aerosol) %>% 
  mutate(Condition = case_when(
    Condition == "no impactor" ~ "none",
    Condition == "5 um impactor" ~ "5 um",
    Condition == "2.5 um impactor" ~ "2 um",
    Condition == "aerosol transmission" ~ "aerosol",
    Condition == "direct transmission" ~ "direct"
  ))


# Add back the stock variants and remove the fixed variants and annotate
plotting.merged.df = amended.merged.df %>% 
  filter(Animal != "stock") %>% 
  left_join(., amino_acid_annotations, by = "SNP") %>% 
  bind_rows(stock.variants) %>% 
  filter(!SNP %in% fixed.in.isolate) %>% 
  mutate(AA_Change = paste(GENE, ":", " ", REF_AA, POS_AA, ALT_AA, sep="")) %>% 
  mutate(AA_Change = case_when(
    is.na(GENE) ~ "intergenic", 
    TRUE ~ AA_Change
  )) %>% 
  mutate(Mutation = paste(SNP, " (", AA_Change , ")")) 
    
```

```{r Plotting Data by Cage, echo=T, warning=F}

cages = plotting.merged.df %>% 
  pull(Experiment) %>% 
  unique()

plots = foreach(i = 1:length(cages)) %do% {
  
  # Get the cage that we're plotting
  cage = cages[i]
    
  # Only plot the SNPs for this cage that are present in more than one time point
  plot.snps = plotting.merged.df %>% 
    filter(Experiment == cage) %>% 
    select(SNP, DPI, Pair) %>% 
    distinct() %>%
    group_by(SNP) %>% 
    count() %>% 
    filter(n > 1) %>% 
    pull(SNP)
  
  # Get the names of the contact animal to exclude stock from these facets
  contacts = plotting.merged.df %>% 
    filter(Experiment == cage & Pair == "contact") %>% 
    pull(Animal) %>% 
    unique()

  # Plot the variants over time
  plt = plotting.merged.df %>% 
    filter(Experiment == cage) %>% 
    filter(SNP %in% plot.snps) %>% 
    filter(!(Pair == "stock" & Animal %in% contacts)) %>% 
    mutate(Grouping = paste(Mutation, Pair, Animal, sep = "-")) %>% 
    ggplot(aes(x = DPI, y = AF, col = Pair, group = Grouping)) +
      geom_point(size = 1.5) + 
      geom_line() + 
      facet_grid(rows = vars(Animal), cols = vars(reorder(Mutation, POS))) + 
      xlab("Days Post Infection") + 
      ylab("Allele Frequency") + 
      ggtitle(paste("Experiment", cage)) + 
      scale_color_manual(values = c("#bd0000", "#0013bd", "#bd00ba")) + 
      scale_y_continuous(breaks = c(0, 0.5, 1)) + 
      theme_bw() + 
      theme(plot.title = element_text(hjust = 0.5)) +                                                               
      theme(strip.text.x = element_text(size = 5)) + 
      theme(legend.position = "bottom")
  
}

```

### No Impactor

Let's look at the cages with no impactors. Are there any interesting patterns here? 

```{r Cage 10.1, fig.align='center', fig.width=16, fig.height=4, message=FALSE}

suppressMessages(print(plots[[1]]))
ggsave(paste(figures.dir, "cage_10.1_no_impactor.png", sep=""), width = 16, height = 4)

```

```{r Cage 10.2, fig.align='center', fig.width=16, fig.height=4, message=FALSE}

suppressMessages(print(plots[[2]]))
ggsave(paste(figures.dir, "cage_10.2_no_impactor.png", sep=""), width = 16, height = 4)

```

### 5um Impactor

Now let's look for the tighter aerosol size of 5um. 

```{r Cage 10.3, fig.align='center', fig.width=16, fig.height=4, message=FALSE}

suppressMessages(print(plots[[3]]))
ggsave(paste(figures.dir, "cage_10.3_5um_impactor.png", sep=""), width = 16, height = 4)

```

```{r Cage 10.4, fig.align='center', fig.width=16, fig.height=4, message=FALSE}

suppressMessages(print(plots[[4]]))
ggsave(paste(figures.dir, "cage_10.4_5um_impactor.png", sep=""), width = 16, height = 4)

```

### 2.5um Impactor

Now let's look for the tighter aerosol size of 2.5um. 


```{r Cage 10.8, fig.align='center', fig.width=16, fig.height=4, message=FALSE}

suppressMessages(print(plots[[5]]))
ggsave(paste(figures.dir, "cage_10.8_2.5um_impactor.png", sep=""), width = 16, height = 4)

```

```{r Cage 10.9, fig.align='center', fig.width=16, fig.height=4, message=FALSE}

suppressMessages(print(plots[[6]]))
ggsave(paste(figures.dir, "cage_10.9_2.5um_impactor.png", sep=""), width = 16, height = 4)

```

### Direct v. Aerosol transmission

Now let's compare direct transmission and aerosol transmission.

```{r Cage 19, fig.align='center', fig.width=16, fig.height=6, message=FALSE}

suppressMessages(print(plots[[7]]))
ggsave(paste(figures.dir, "cage_19_aerosol_v_direct.png", sep=""), width = 16, height = 6)

```

## Summarize Variants Over Time

It is interesting to see what happens to specific variants over time. I'm going to use the variants that we identified in the filtered dataset (the variants used to calculate the bottleneck). This means that variants are present in more than one sequencing run, found at more than 2% AF, found with more an 25 runs with alternative alleles, and less then 90% strand bias. In addition, these variants are not found in primer sequences. 

One additional thing I'll do is focus on variants that are found in more than one time point (in either the donor or the contact). These are the most interesting variants for observing patterns over time. 

```{r Summarized Variants, fig.align='center', fig.width=8, fig.height=6, message=FALSE}

# Only plot the SNPs in both rounds of sequencing (same as bottleneck)
snps.in.both.sequencing.rounds = filtered.merged.df %>% 
  select(SNP, Sequencing_Round) %>% 
  distinct() %>% 
  group_by(SNP) %>% 
  count() %>% 
  rename(SeqRoundObsv = n) %>% 
  filter(SeqRoundObsv > 1) %>% 
  pull(SNP)

# Only plot the SNPs with more than one time point per animal 
snps.in.multiple.timepoints = filtered.merged.df %>% 
  select(SNP, Animal, DPI) %>% 
  group_by(SNP, Animal) %>% 
  count()  %>% 
  rename(DPIObsv = n) %>% 
  filter(DPIObsv > 1) %>% 
  pull(SNP)
  
# Only plot the SNPs also found in the isolate (stock)
snps.to.plot = plotting.merged.df %>% 
  filter((SNP %in% snps.in.both.sequencing.rounds) & (SNP %in% snps.in.multiple.timepoints)) %>% 
  select(SNP, Pair) %>% 
  distinct() %>% 
  mutate(Tally = 1) %>% 
  pivot_wider(names_from = "Pair", values_from = "Tally", values_fill = 0) %>% 
  filter(stock != 0) %>% 
  pull(SNP)

# Plot the SNPs using the same filters as the bottleneck
plotting.merged.df %>% 
  filter(SNP %in% snps.to.plot) %>% 
  mutate(Grouping = paste(Mutation, Pair, Animal, sep = "-")) %>% 
  ggplot(aes(x = DPI, y = AF, col = Pair, group = Grouping)) +
    geom_point(size = 3.5, alpha = 0.5) + 
    geom_line(alpha = 0.5) + 
    facet_wrap(~fct_reorder(Mutation, POS, min)) +
    xlab("Days After Inital Donor Infection") + 
    ylab("Allele Frequency") +
    scale_color_manual(values = c("#bd00ba","#0013bd", "#bd0000"),
                       breaks =c("stock", "donor", "contact")) + 
    scale_y_continuous(breaks = c(0, 0.5, 1)) + 
    theme_bw(18) + 
    theme(plot.title = element_text(hjust = 0.5)) +                                                           
    theme(legend.position = "bottom")
  
ggsave(paste(figures.dir, "variants-over-time-and-condition.png", sep=""), width = 12, height = 10)

```

I'll also make a version of this plot that show the other contact and donor samples even if a SNP wasn't observed in those animals. If a SNP wasn't observed, it's allele frequency is set to 0. 

```{r Summarized Variants All SNPs , fig.align='center', fig.width=8, fig.height=6, message=FALSE}

# Name of each facet title
snp.info = plotting.merged.df %>% 
  filter(SNP %in% snps.to.plot) %>% 
  select(SNP, Mutation, POS) %>% 
  distinct()

# Get the Cartesian project of all snps to plot and conditions
sample.snp.cartesian.product = amended.merged.df %>% 
  filter(Animal != 'stock') %>% 
  select(Animal, DPI, Pair) %>% 
  distinct() %>%
  merge(data.frame(SNP = snps.to.plot), by = NULL)

# Format the plotting data frame, adding in the unobserved SNPs as AF = 0
plotting.merged.df %>% 
  filter(SNP %in% snps.to.plot) %>% 
  select(Animal, Pair, DPI, SNP, AF) %>% 
  full_join(., sample.snp.cartesian.product) %>% 
  mutate(AF = if_else(is.na(AF), 0, AF)) %>% 
  distinct() %>% 
  left_join(., snp.info) %>% 
  mutate(Grouping = paste(Mutation, Pair, Animal, sep = "-")) %>% 
  ggplot(aes(x = DPI, y = AF, col = Pair, group = Grouping)) +
    geom_point(size = 3.5, alpha = 0.5) + 
    geom_line(alpha = 0.5) + 
    facet_wrap(~fct_reorder(Mutation, POS, min)) +
    xlab("Days After Inital Donor Infection") + 
    ylab("Allele Frequency") +
    scale_color_manual(values = c("#bd00ba","#0013bd", "#bd0000"),
                       breaks =c("stock", "donor", "contact")) + 
    scale_y_continuous(breaks = c(0, 0.5, 1)) + 
    theme_bw(18) + 
    theme(plot.title = element_text(hjust = 0.5)) +                                                           
    theme(legend.position = "bottom")
  
ggsave(paste(figures.dir, "variants-over-time-and-condition-all-snps.png", sep=""), width = 12, height = 10)

```

```{r Summarized Variants All SNPs , fig.align='center', fig.width=20, fig.height=5, message=FALSE}

# Get the Cartesian project of all snps to plot and conditions
sample.snp.cartesian.product = amended.merged.df %>% 
  filter(Animal != 'stock') %>% 
  select(Animal, DPI, Pair, Experiment) %>% 
  distinct() %>%
  merge(data.frame(SNP = snps.to.plot), by = NULL)

# Format the plotting data frame, adding in the unobserved SNPs as AF = 0
plotting.merged.df %>% 
  filter(SNP %in% snps.to.plot) %>% 
  select(Animal, Pair, Experiment, DPI, SNP, AF) %>% 
  full_join(., sample.snp.cartesian.product) %>% 
  mutate(AF = if_else(is.na(AF), 0, AF),
         Experimental_Condition = if_else(Experiment == 19.0, "Aerosol v. Direct", "Impactor")
         ) %>% 
  distinct() %>% 
  left_join(., snp.info) %>% 
  mutate(Grouping = paste(Mutation, Pair, Animal, sep = "-")) %>% 
  ggplot(aes(x = DPI, y = AF, col = Pair, group = Grouping)) +
    geom_point(size = 3.5, alpha = 0.5) + 
    geom_line(alpha = 0.5) + 
    facet_grid(cols = vars(fct_reorder(Mutation, POS, min)), rows = vars(Experimental_Condition)) +
    xlab("Days After Inital Donor Infection") + 
    ylab("Allele Frequency") +
    scale_color_manual(values = c("#bd00ba","#0013bd", "#bd0000"),
                       breaks =c("stock", "donor", "contact")) + 
    scale_y_continuous(breaks = c(0, 0.5, 1)) + 
    scale_x_continuous(breaks = c(0, 2, 4, 6, 8, 10)) + 
    theme_bw(18) + 
    theme(plot.title = element_text(hjust = 0.5)) +                                                           
    theme(legend.position = "bottom") +
    theme(strip.text.x = element_text(size = 12))
  
ggsave(paste(figures.dir, "split-variants-over-time-and-condition-all-snps.png", sep=""), width = 22, height = 6)

```

Finally, I'll try a version where I split the variants over time based on the two experiments (Impactor and Aerosol v. Direct) and change the shape of the points based on the specific experimental condition (impactor size, aerosol, or direct).

```{r Split Summarized Variants All SNPs , fig.align='center', fig.width=10, fig.height=5, message=FALSE}

# Get the Cartesian project of all snps to plot and conditions
sample.snp.cartesian.product = amended.merged.df %>% 
  filter(Animal != 'stock') %>% 
  select(Animal, DPI, Pair, Experiment, Condition) %>% 
  distinct() %>%
  merge(data.frame(SNP = snps.to.plot), by = NULL)


# Format the plotting data frame, adding in the unobserved SNPs as AF = 0

shapes = c(19, 17, 15, 23)
names(shapes) = c("stock", "No Impactor", "2.5 um Impactor", "5 um Impactor")

plotting.merged.df %>% 
  filter(SNP %in% snps.to.plot) %>% 
  select(Animal, Pair, Experiment, Condition, DPI, SNP, AF) %>% 
  full_join(., sample.snp.cartesian.product) %>% 
  mutate(AF = if_else(is.na(AF), 0, AF),
         Experimental_Condition = if_else(Experiment == 19.0, "Aerosol v. Direct", "Impactor")
         ) %>% 
  distinct() %>% 
  left_join(., snp.info) %>% 
  mutate(Grouping = paste(Mutation, Pair, Animal, sep = "-")) %>% 
  filter(Experimental_Condition == "Impactor") %>% 
  mutate(Impactor = case_when(
    Condition == "none" ~ "No Impactor",
    Condition == "5 um" ~ "5 um Impactor",
    Condition == "2 um" ~ "2.5 um Impactor",
     Condition == "isolate stock" ~ "stock"
  )) %>%  
  ggplot(aes(x = DPI, y = AF, col = Pair, fill = Pair, group = Grouping, shape = Impactor)) +
    geom_point(size = 3.5, alpha = 0.5) + 
    geom_line(alpha = 0.5) + 
    facet_wrap(~fct_reorder(Mutation, POS, min), nrow = 2) +
    xlab("Days After Inital Donor Infection") + 
    ylab("Allele Frequency") +
    scale_color_manual(values = c("#bd00ba","#0013bd", "#bd0000"),
                       breaks =c("stock", "donor", "contact")) + 
    scale_fill_manual(values = c("#bd00ba","#0013bd", "#bd0000"),
                     breaks =c("stock", "donor", "contact")) + 
    scale_y_continuous(breaks = c(0, 0.5, 1)) + 
    scale_x_continuous(breaks = c(0, 2, 4, 6, 8, 10)) + 
    scale_shape_manual(values = shapes) + 
    theme_bw(16) + 
    theme(plot.title = element_text(hjust = 0.5)) +                                                           
    theme(legend.position = "bottom") +
    theme(strip.text.x = element_text(size = 12))
  
ggsave(paste(figures.dir, "figure_impactor_variants_over_time.png", sep=""), width = 12, height = 6)

shapes = c(19, 17, 15)
names(shapes) = c("stock", "Aerosol", "Direct")

plotting.merged.df %>% 
  filter(SNP %in% snps.to.plot) %>% 
  select(Animal, Pair, Experiment, Condition, DPI, SNP, AF) %>% 
  full_join(., sample.snp.cartesian.product) %>% 
  mutate(AF = if_else(is.na(AF), 0, AF),
         Experimental_Condition = if_else(Experiment == 19.0, "Aerosol v. Direct", "Impactor")
         ) %>% 
  distinct() %>% 
  left_join(., snp.info) %>% 
  mutate(Grouping = paste(Mutation, Pair, Animal, sep = "-")) %>% 
  filter(Experimental_Condition == "Aerosol v. Direct") %>% 
  mutate(Impactor = case_when(
    Condition == "aerosol" ~ "Aerosol",
    Condition == "direct" ~ "Direct",
    Condition == "isolate stock" ~ "stock"
  )) %>%  
  ggplot(aes(x = DPI, y = AF, col = Pair, fill = Pair, group = Grouping, shape = Impactor)) +
    geom_point(size = 3.5, alpha = 0.5) + 
    geom_line(alpha = 0.5) + 
    facet_wrap(~fct_reorder(Mutation, POS, min), nrow = 2) +
    xlab("Days After Inital Donor Infection") + 
    ylab("Allele Frequency") +
    scale_color_manual(values = c("#bd00ba","#0013bd", "#bd0000"),
                       breaks =c("stock", "donor", "contact")) + 
    scale_fill_manual(values = c("#bd00ba","#0013bd", "#bd0000"),
                     breaks =c("stock", "donor", "contact")) + 
    scale_y_continuous(breaks = c(0, 0.5, 1)) + 
    scale_x_continuous(breaks = c(0, 2, 4, 6, 8, 10)) + 
    scale_shape_manual(values = shapes) + 
    theme_bw(16) + 
    theme(plot.title = element_text(hjust = 0.5)) +                                                           
    theme(legend.position = "bottom") +
    theme(strip.text.x = element_text(size = 12))
  
ggsave(paste(figures.dir, "figure_aerosol_v_direct_variants_over_time.png", sep=""), width = 12, height = 6)

```

## Key Takeaways 

There are several key takeaways from this notebook: 

1. There is a mutation in the Spike multi-basic cleavage site (`S:R682L`) that is never transmitted, but it's high frequency in the stock and much lower frequency in the contact. 
2. There is a mutation in Spike (`S:H655Y`) that exhibits behavior that's consistent with arising multiple times convergently. Similar to (`S:N74K`). 
3. However, what's a little puzzling is that no mutations appear to fix over time. 

